File: Published/SourceCode/Python/Other/Hashcode.py

class Person:
    def __init__(self, first, last):
        self.firstName = first
        self.lastName = last

    def __hash__(self):
        # Note: The function ``hash(x)`` calls ``x.__hash__()``
        return hash(self.firstName) + hash(self.lastName)

    def __str__(self):
        return self.firstName + " " + self.lastName

if __name__ == '__main__':
    import sys
    someone = Person(sys.argv[1], sys.argv[2])
    print(f"The hash code for {someone} is {hash(someone)}")

    def hashInt(x):
        return x % 16

    def hashString(s, M):
        sum = 0
        for c in s:
            sum += ord(c)
        return sum % M

    def hashStringImproved(s, M):
        sum = 0
        for c in s:
            sum = 31 * sum + ord(c)
        return sum % M

File: Published/SourceCode/Java/Other/Hashcode.java


class Person {
    public String firstName;
    public String lastName;

    Person(String first, String last) {
        firstName = first;
        lastName = last;
    }

    public String toString() {
        return firstName + " " + lastName;
    }

    public int hashCode() {
        return firstName.hashCode() + lastName.hashCode();
    }
}


class Hashcode {

    int hashInt(int x) {
        return x % 16;
    }

    int hashString(String s, int M) {
        int sum = 0;
        for (char c : s.toCharArray())
            sum += c;
        return sum % M;
    }
  
    int hashStringImproved(String s, int M) {
        int sum = 0;
        for (char c : s.toCharArray())
            sum = 31 * sum + c;
        return sum % M;
    }

    public static void main(String[] args) {
        if (args.length == 0)
            args = new String[] {"Peter", "Ljunglöf"};
        Person someone = new Person(args[0], args[1]);
        System.out.println("The hash code for " + someone + " is " + someone.hashCode());
    }
}
